File: Virtual Machine/Nova.VirtualMachine/Nova.VirtualMachine/RuntimeContext.cpp
#include "RuntimeContext.h"
#include "NovFile.h"
#include "Exec.h"
#include "RuntimeStruct.h"
#include "Call.h"

Null* RuntimeContext::NULL_VALUE = new Null(); /* Should we create a class to store this kinda types? FALSE , TRUE ? */


RuntimeContext::RuntimeContext(NovFile* file)
{
	this->novFile = file;
}

void RuntimeContext::Initialize()
{
	for (ByteClass* byteClass : this->novFile->byteClasses)
	{
		for (ByteField* field : byteClass->fields)
		{
			field->Initializer(this);
		}
	}
}



RuntimeContext::RuntimeElement RuntimeContext::GetConstant(int id)
{
	return this->novFile->constants[id];
}

void RuntimeContext::PushStack(RuntimeContext::RuntimeElement  element)
{
	stack.emplace_back(element);
}

RuntimeContext::RuntimeElement RuntimeContext::PopStack()
{
	size_t stackSize = stack.size();
	RuntimeContext::RuntimeElement value = stack.at(stackSize - 1);
	stack.erase(stack.begin() + stackSize - 1);
	return value;
}

RuntimeContext::RuntimeElement RuntimeContext::StackMinus(int minus)
{
	return this->stack[this->stack.size() - 1 - minus];
}

size_t RuntimeContext::GetStackSize()
{
	return this->stack.size();
}

RuntimeStruct* RuntimeContext::CreateStruct(int classId)
{
	return new RuntimeStruct(novFile->byteClasses[classId]);
}

RuntimeStruct* RuntimeContext::GetCurrentStruct()
{
	return structsStack.at(structsStack.size() - 1);
}

RuntimeContext::RuntimeElement RuntimeContext::Get(int classId, int fieldId)
{
	return this->novFile->byteClasses[classId]->fields[fieldId]->value;
}

void RuntimeContext::Set(int classId, int fieldId, RuntimeContext::RuntimeElement value)
{
	this->novFile->byteClasses[classId]->fields[fieldId]->value = value;
}

File: Compiler/Nova.VirtualMachine/Runtime/RuntimeContext.cs
using Nova.VirtualMachine.IO;
using Nova.VirtualMachine.Members;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Nova.VirtualMachine.Runtime
{
    public class RuntimeContext
    {
        public static Null NULL_VALUE = new Null();

        public NovFile NovFile
        {
            get;
            set;
        }
        private List<object> Stack
        {
            get;
            set;
        }
        private List<RuntimeStruct> StructStack
        {
            get;
            set;
        }
        private List<ByteMethod> CallStack
        {
            get;
            set;
        }
        public RuntimeContext(NovFile file)
        {
            this.NovFile = file;
            this.Stack = new List<object>();
            this.StructStack = new List<RuntimeStruct>();
            this.CallStack = new List<ByteMethod>();
        }

        public void Initialize()
        {
            foreach (var byteClass in NovFile.Classes)
            {
                foreach (var field in byteClass.Fields)
                {
                    field.Initializer(this);
                }
            }

        }
        public void CallMain()
        {
            ByteMethod method = this.NovFile.GetMainMethod();
            Call(method);
        }

        public object GetConstant(int index)
        {
            return this.GetExecutingClass().Constants[index];
        }

        private ByteClass GetExecutingClass()
        {
            return CallStack[CallStack.Count - 1].Parent;
        }
        public void Call(int classId, int methodId)
        {
            ByteMethod method = this.NovFile.Classes[classId].Methods[methodId];
            Call(method);
        }


        public void Call(RuntimeStruct st, int methodId)
        {
            StructStack.Add(st);

            ByteMethod method = st.TypeClass.Methods[methodId];
            Call(method);

            StructStack.RemoveAt(StructStack.Count - 1);
        }

        private void Call(ByteMethod method)
        {
            CallStack.Add(method); // we push call stack

            object[] locales = new object[method.ParametersCount];

            for (int i = method.ParametersCount - 1; i >= 0; i--)
            {
                locales[i] = PopStack();
            }

            Exec.Execute(this, locales, method.Block.Instructions);

            CallStack.RemoveAt(CallStack.Count - 1);
        }

        public int GetStackSize()
        {
            return Stack.Count;
        }

        public object PopStack()
        {
            int size = GetStackSize();
            object value = Stack[size - 1];
            Stack.RemoveAt(size - 1);
            return value;
        }
        public void PushStack(object value)
        {
            Stack.Add(value);
        }
    }
}
